src/iso15118/io/sdp_server.cpp

```diff
@@ -120,6 +123,70 @@ void SdpServer::send_response(const PeerRequestContext& request, const Ipv6EndPo
 
     sendto(fd, v2g_packet, sizeof(v2g_packet), 0, reinterpret_cast<const sockaddr*>(&request.address), peer_addr_len);
 }
+TlsKeyLoggingServer::TlsKeyLoggingServer(const std::string& interface_name, uint16_t port) {
+    static constexpr auto LINK_LOCAL_MULTICAST = "ff02::1";
+
+    fd = socket(AF_INET6, SOCK_DGRAM, 0);
+    if (fd < 0) {
+        const auto error_msg = adding_err_msg("Could not create socket");
+        log_and_throw(error_msg.c_str()); // FIXME(sl): Find better handling
+    }
+
+    // source setup
+
+    // find port between 49152-65535
+    auto could_bind = false;
+    auto source_port = 49152;
+    for (; source_port < 65535; source_port++) {
+        sockaddr_in6 source_address = {AF_INET6, htons(source_port)};
+        if (bind(fd, reinterpret_cast<sockaddr*>(&source_address), sizeof(sockaddr_in6)) == 0) {
+            could_bind = true;
+            break;
+        }
+    }
+
+    if (!could_bind) {
+        const auto error_msg = adding_err_msg("Could not bind");
+        log_and_throw(error_msg.c_str());
+    }
+
+    logf_info("UDP socket bound to source port: %u", source_port);
+
+    const auto index = if_nametoindex(interface_name.c_str());
+    auto mreq = ipv6_mreq{};
+    mreq.ipv6mr_interface = index;
+    if (inet_pton(AF_INET6, LINK_LOCAL_MULTICAST, &mreq.ipv6mr_multiaddr) <= 0) {
+        const auto error_msg = adding_err_msg("Failed to setup multicast address");
+        log_and_throw(error_msg.c_str());
+    }
+    if (setsockopt(fd, IPPROTO_IPV6, IPV6_ADD_MEMBERSHIP, &mreq, sizeof(mreq)) < 0) {
+        const auto error_msg = adding_err_msg("Could not add multicast group membership");
+        log_and_throw(error_msg.c_str());
+    }
+
+    if (setsockopt(fd, IPPROTO_IPV6, IPV6_MULTICAST_IF, &index, sizeof(index)) < 0) {
+        const auto error_msg = adding_err_msg("Could not set interface name:" + std::string(interface_name));
+        log_and_throw(error_msg.c_str());
+    }
+
+    destination_address = {AF_INET6, htons(port)};
+    if (inet_pton(AF_INET6, LINK_LOCAL_MULTICAST, &destination_address.sin6_addr) <= 0) {
+        const auto error_msg = adding_err_msg("Failed to setup server address, reset key_log_fd");
+        log_and_throw(error_msg.c_str());
+    }
+}
+
+TlsKeyLoggingServer::~TlsKeyLoggingServer() {
+    logf_info("Shutting down TlsKeyLoggingServer server!");
+    if (fd != -1) {
+        close(fd);
+    }
+}
```

Comment:
Keep in mind that an `fd` is a unique resource and currently the `TlsKeyLoggingServer` is copy-assignable/copy-constructable, meaning it could miss to close an the `fd` if it is assigned a new one, and it could be closed two times.  Of course you'll probably will not copy the `TlsKeyLoggingServer` intentionally but in general it is a good idea to just not allow it (like it is with `std::unique_ptr`).  I know of three possible solutions to this problem:
 - use a std::unique_ptr<int> to wrap the fd, and as soon as you have a move-only member variable, the containing class itself will become move-only
 - use a custom move-only class which wraps an fd
 - delete the copy constructors of `TlsKeyLoggingServer`
 
 the latter two possibilities need some care because it is easy to miss some the move/copy constructors.  Also keep in mind that the move constructor needs to set the fd to a special value or state, so that the constructor can infer if it needs to `close` it or not. In case of `std::unique_ptr` it will be `nullptr` after move.

Reply:
I understand your concerns, but isn't this duplicated in this case? Since a `unique_ptr` is already defined in the `SSLContext` for the whole `TlsKeyLoggingServer` object. In my opinion, `fd` can never or not twice be closed by this  `key_server` object in `SSLContext`. 

If, for example, `fd` is wrapped here with a `unique_ptr,` this should also be done for the `fd's` in the other connection_* classes. However, it is also the case that the whole `ConnectionSSL` or `ConnectionPlain` are defined as `unique_ptr`.